### DISCUSSION TITLE
check for schema mismatch inside sqlalchemy orm context manager

### DISCUSSION THREAD

**Discussion by yishang11** [sqlalchemy/sqlalchemy#11000] [Usage Questions]
I am using sqlalchemy 1.4 with orm.
I need to support many sqlite databases created using orm API over time. As the orm classes change the database schema evolved over time. I created a shared Alembic environment describing schema changes so I could upgrade db having outdated schema.

Before any orm operations on the db, I would copy over the shared Alembic environment and check the db's schema, and then upgrade if needed. But to further prevent schema related errors, I wanted to know if I could also do a quick schema check within the orm context manager I created, and raise an error when a non-empty db has schema mismatch?
Currently if I just use the orm context manager to open a database with old schema 
- If the current schema has tables added, then the context would add tables silently, changing the schema, because I have `Base.metadata.create_all()` in my `__enter__()`.
- If the current schema has a column deleted from a table, querying that table won't fail, and the schema has not changed.
- If the current schema has a column added to a table, then querying the table would fail with `OperationalError`.

I basically want the orm context manager to fail loudly for a mismatched schema, but this check should be cheap enough to put inside the `__init__()` method. Any suggestions/comments would be appreciated. Thank you!

**Comment by zzzeek** [comment]
do you want to check the schema by reflecting the whole DB and comparing to MetaData?  that's not what I'd call "cheap".  The other way is to look in alembic_version and make sure it's up to date using https://alembic.sqlalchemy.org/en/latest/cookbook.html#test-current-database-revision-is-at-head-s .     that would be more cheap.

other than those two I dont know another approach
